packages/drawer/src/ux-drawer.ts

```diff
@@ -0,0 +1,315 @@
+        this.setViewportType();
+        this.handlingEvent = false;
+      });
+    } else {
+      setTimeout(() => {
+        this.setViewportType();
+        this.handlingEvent = false;
+      }, 100);
+    }
+    this.setViewportType();
+  }
+
+  public themeChanged(newValue: any) {
+    if (newValue != null && newValue.themeKey == null) {
+      newValue.themeKey = 'drawer';
+    }
+
+    this.styleEngine.applyTheme(newValue, this.element);
+  }
+
+  public setViewportType() {
+    this.viewportType = window.innerWidth < this.modalBreakpoint ? 'mobile' : 'desktop';
+  }
+
+  @computedFrom('type', 'viewportType')
+  public get computedType(): 'standard' | 'modal' {
+    return this.viewportType === 'mobile' ? 'modal' : this.type;
+  }
+
+  public overlayClick(event: Event): any {
+    for (let element of event.composedPath() ) {
+      if (element === this.contentElement) {
+        return true; // this allow normal behvior when clicking on elements inside the drawer
+      }
+    }
+    if (!this.overlayDismiss) {
+      event.stopPropagation();
+      return;
+    }
+    this.dismiss();
+  }
+
+  public async dismiss(event?: Event) {
+    if (event) {
+      event.stopPropagation();
+    }
+    const result: ModalServiceResult = {
+      wasCancelled: true,
+      output: undefined
+    };
+    if (!await this.prepareClosing(result)) {
+      return;
+    }
+    await this.hide();
+    const dismissEvent = DOM.createCustomEvent('dismiss', {bubbles: true});
+    this.element.dispatchEvent(dismissEvent);
+  }
+
+  public async ok(output?: any, event?: Event) {
+    if (event) {
+      event.stopPropagation();
+    }
+    const result: ModalServiceResult = {
+      wasCancelled: false,
+      output
+    };
+    if (!await this.prepareClosing(result)) {
+      return;
+    }
+    await this.hide();
+    const okEvent = DOM.createCustomEvent('ok', {bubbles: true, detail: result.output});
+    this.element.dispatchEvent(okEvent);
+  }
+
+  private async prepareClosing(result: ModalServiceResult): Promise<boolean> {
+    const layer = this.modalService.getLayer(this);
+    if (layer) {
+      if (!await this.modalService.callCanDeactivate(layer, result)) {
+        return false;
+      }
+      try {
+        await this.modalService.callDetached(layer);
+        await this.modalService.callDeactivate(layer, result);
+      } catch (error) {
+        log.error(error);
+      } 
+    }
+    return true;
+  }
+
+  public stop(event: Event) {
+    event.stopPropagation();
+  }
+
+  private getAnimationDuration() {
+    const overlayElementDuration: string = window.getComputedStyle(this.overlayElement).transitionDuration || '';
+    const contentDuration: string = window.getComputedStyle(this.contentElement).transitionDuration || '';
+    // overlayElementDuration and contentDuration are string like '0.25s'
+    return Math.max(parseFloat(overlayElementDuration), parseFloat(contentDuration)) * 1000;
```

Comment:
I left the following comment in the code:

```ts
// In order to allow precise animation we allow different duration
// value for animating the overlay and the drawer. In most cases it will
// be the same value but we can imagine a fast overlay and slower modal
// apearence for exemple
// Because the duration is used to determine when we can safely assume the
// modal disappeard, we only keep the maximum value.
```